Provenance v1: add verification section, clarify model (#618)
slsa-framework/slsa: docs/provenance/v1.md

[Verification]: #verification

Verification of provenance encompasses the following steps.

![Threats covered by each step](../images/supply-chain-threats-build-verification.svg)

### Step 1: Check SLSA Build level

[verify-step-1]: #check-slsa-build-level

First, check the SLSA Build level by comparing the artifact to its provenance
and the provenance to a preconfigured root of trust. The goal is to ensure that
the provenance actually applies to the artifact in question and to assess the
trustworthiness of the provenance. This mitigates some or all of [threats] "D",
"F", "G", and "H", depending on SLSA Build level and where verification happens.

Up front:

> -   Establish roots of trust, meaning the recognized pairs of builder public key

Reviewer: Nit: Should we _identify_, rather than _establish_, roots of trust?

Author: What about *configure*?

Reviewer: WFM!

    identity + `builder.id`, mapped to SLSA Build levels. Different
    organizations might use different roots of trust, but usually a verifier

>     uses the same root of trust for all packages.

Reviewer: It might be useful to provide a pedagogical example of what this might look like, in the form of a JSON snippet perhaps?

Author: FYI, my previous commit said "max level" but I realize now it's just "level". There are no futher checks that could cause it to be downgraded.

    persistence between builds or theft of the provenance signing key. In other
    words, SLSA Build L3 establishes that the provenance is accurate and
    trustworthy, assuming you trust the build platform.
    -   IMPORTANT: SLSA Build L3 does **not** cover compromise of the build
        platform itself, such as by a malicious insider. Instead, verifiers
        SHOULD carefully consider which build platforms are added to the roots
        of trust. For advice on establishing trust in build platforms, see
        [Verifying build systems](/spec/v1.0/verifying-systems).
-   [Threat "F"]: SLSA Build L2 covers tampering of the artifact or provenance
    after the build. This is accomplished by verifying the `subject` and
    signature in the steps above.
-   [Threat "G"]: Verification by the consumer or otherwise outside of the
    package registry covers compromise of the registry itself. (Verifying within
    the registry at publication time is also valuable, but does not cover Threat
    "G" or "H".)
-   [Threat "H"]: Verification by the consumer covers compromise of the package
    in transit. (Many ecosystems already address this threat using package

>     signatures.)

Reviewer: Is it worth adding:
```suggestion
    signatures or checksums.)
```
?

In many open source ecosystems signatures are not particularly common, but checksums/hash-checking are increasingly common. Is it worth mentioning those here?

    -   NOTE: SLSA does not cover adversaries tricking a consumer to use an
        unintended package, such as through typosquatting.

[Threat "D"]: /spec/v1.0/threats#d-compromise-build-process
[Threat "F"]: /spec/v1.0/threats#f-upload-modified-package
[Threat "G"]: /spec/v1.0/threats#g-compromise-package-repo
[Threat "H"]: /spec/v1.0/threats#h-use-compromised-package

[processing-model]: https://github.com/in-toto/attestation/tree/main/spec#processing-model

### Step 2: Check expectations

[verify-step-2]: #check-expectations

Next, check that the package's provenance meets expectations for that package in
order to mitigate [threat "C"].

In our threat model, the adversary has ability to invoke a build and to publish
to the registry but not to write to the source repository, nor do they have
insider access to any trusted systems. Expectations MUST be sufficient to detect
or prevent this adversary from injecting unofficial behavior into the package.
Example threats in this category include building from an unofficial fork or
abusing a build parameter to modify the build. Usually expectations identify the

> canonical source repository (which is the main external parameter) and any other

Reviewer: nit: I think this a bit more explicit/easier to understand if we use the field name:

```suggestion
canonical source repository (which is the main entry in `externalParameters`) and any other
```

security-relevant external parameters.

The expectations SHOULD cover the following:

| What | Why |
| ---- | --- |
| Builder identity from [Step 1](verify-step-1) | To prevent an adversary from building the correct code on an unintended system |
| `buildType` | To ensure that `externalParameters` are interpreted as intended |
| `externalParameters` | To prevent an adversary from injecting unofficial behavior |

> Verifiers SHOULD reject unrecognized external parameters. It is acceptable to

Reviewer: ```suggestion
Verifiers SHOULD reject unrecognized `externalParameters`. It is acceptable to
```

allow a parameter to have a range of values (possibly any value) if it is known
that any value in the range is safe. Implementations need not special-case the
`buildType` if JSON comparisons are sufficient.

> Possible models (not exhaustive):

Reviewer: This could do with some more context to help it flow, otherwise the jump from verifying expectations to setting expectations is a bit jarring.
```suggestion
Possible models for implementing expectation setting in package ecosystems (not exhaustive):
```

    location.

TIP: Difficulty in setting meaningful expectations for `externalParameters` can
be a sign that the `buildType`'s level of abstraction is too low. For example,
`externalParameters` that record a list of commands to run is likely impractical
to verify because the commands change on every build. Instead, consider a
`buildType` that defines the list of commands in a configuration file in a
source repository, then make the source repository the external parameter. Such
a design is easier to verify because the source repository is constant across
builds.

[Threat "C"]: /spec/v1.0/threats#c-build-from-modified-source

### Step 3: Check dependencies (recursively)

[verify-step-3]: #step-3-check-dependencies-recursively

Finally, recursively check the `resolvedDependencies` as available and to the
extent desired. This mitigates [threat "E"]. While SLSA v1.0 does not have any
requirements on the completeness or verification of `resolvedDependencies`, one
might wish to go beyond SLSA's minimum requirements in order to protect against
threats further up the supply chain.

One possible approach is to recursively [verify][Verification] each entry in

> `resolvedDependencies`. A [Verification Summary Attestation (VSA)][VSA] can be

Reviewer: ```suggestion
`resolvedDependencies`. A [Verification Summary Attestation (VSA)][VSA] can
```
